source/common/config/delta_subscription_impl.h

```diff
@@ -0,0 +1,223 @@
+                   ResourceNameDiff>(std::move(async_client), service_method, random, dispatcher,
+                                     scope, rate_limit_settings),
+        type_url_(Grpc::Common::typeUrl(ResourceType().GetDescriptor()->full_name())),
+        local_info_(local_info), stats_(stats) {
+    request_.set_type_url(type_url_);
+    request_.mutable_node()->MergeFrom(local_info_.node());
+  }
+
+  // Enqueues and attempts to send a discovery request, (un)subscribing to resources missing from /
+  // added to the passed 'resources' argument, relative to resources_. Updates resources_ to
+  // 'resources'.
+  void buildAndQueueDiscoveryRequest(const std::vector<std::string>& resources) {
+    ResourceNameDiff diff;
+    std::set_difference(resources.begin(), resources.end(), resource_names_.begin(),
+                        resource_names_.end(), std::inserter(diff.added_, diff.added_.begin()));
+    std::set_difference(resource_names_.begin(), resource_names_.end(), resources.begin(),
+                        resources.end(), std::inserter(diff.removed_, diff.removed_.begin()));
+
+    for (const auto& added : diff.added_) {
+      resources_[added] = "0";
+      resource_names_.insert(added);
+    }
+    for (const auto& removed : diff.removed_) {
+      resources_.erase(removed);
+      resource_names_.erase(removed);
+    }
+    queueDiscoveryRequest(diff);
+  }
+
+  void sendDiscoveryRequest(const ResourceNameDiff& diff) override {
+    if (!grpcStreamAvailable()) {
+      // Don't immediately try to reconnect: we rely on retry_timer_ for that.
+      ENVOY_LOG(debug, "No stream available to sendDiscoveryRequest for {}", type_url_);
+      return;
+    }
+    if (paused_) {
+      ENVOY_LOG(trace, "API {} paused during sendDiscoveryRequest().", type_url_);
+      return;
+    }
+
+    request_.clear_resource_names_subscribe();
+    request_.clear_resource_names_unsubscribe();
+    std::copy(
+        diff.added_.begin(), diff.added_.end(),
+        ::google::protobuf::RepeatedFieldBackInserter(request_.mutable_resource_names_subscribe()));
+    std::copy(diff.removed_.begin(), diff.removed_.end(),
+              ::google::protobuf::RepeatedFieldBackInserter(
+                  request_.mutable_resource_names_unsubscribe()));
+
+    ENVOY_LOG(trace, "Sending DiscoveryRequest for {}: {}", type_url_, request_.DebugString());
+    sendMessage(request_);
+    request_.clear_error_detail();
+    request_.clear_initial_resource_versions();
+  }
+
+  void subscribe(const std::vector<std::string>& resources) {
+    ENVOY_LOG(debug, "delta subscribe for " + type_url_);
+
+    // Lazily kick off the requests based on first subscription. This has the
+    // convenient side-effect that we order messages on the channel based on
+    // Envoy's internal dependency ordering.
+    if (!subscribed_) {
+
+      subscribed_ = true;
+    }
+    buildAndQueueDiscoveryRequest(resources);
+  }
+
+  void pause() {
```

Comment:
There's a bunch of tricky stuff going on with pause/resume right now in the context of CDS, see https://github.com/envoyproxy/envoy/issues/5174. Do we need to do anything to ensure incremental is future proof for this?

Reply:
I think the approach currently being taken there - Envoy should not ACK CDS until the cluster is actually ready (and xDS servers should understand that ACKs are a meaningful part of ADS' dependency mechanism) - makes sense. Should be easy enough to add, since the pause/resume that implementation uses is also present here.